src/java/org/apache/cassandra/replication/PendingLocalTransfer.java

```diff
@@ -0,0 +1,207 @@
+                Preconditions.checkState(range.contains(bounds.left));
+                Preconditions.checkState(range.contains(bounds.right));
+            }
+        }
+
+        Preconditions.checkNotNull(range);
+        return range;
+    }
+
+    private boolean isFullReplica()
+    {
+        ClusterMetadata cm = ClusterMetadata.current();
+        Keyspace ks = Keyspace.open(keyspace);
+        ReplicaGroups writes = cm.placements.get(ks.getMetadata().params.replication).writes;
+        EndpointsForRange replicas = writes.forRange(range.right).get();
+        return replicas.containsSelf() && replicas.selfIfPresent().isFull();
+    }
+
+    /**
+     * Safely move a transfer into the live set. This must be crash-safe, and the primary invariant we need to
+     * preserve is a transfer is only added to the live set iff the transfer ID is present in its mutation summaries.
+     *
+     * We don't validate checksums here, mostly because a transfer can be activated during a read, if one replica
+     * missed the TransferActivation. Transfers should not be pending for very long, and should be protected by
+     * internode integrity checks provided by TLS.
+     *
+     * TODO: Clear out the row cache and counter cache, like {@link CassandraStreamReceiver#finished}.
+     * TODO: Don't add to the live set if coordinator and not an owner for the range
+     */
+    public void activate(TransferActivation activation)
+    {
+        if (activated)
+            return;
+
+        Preconditions.checkState(isFullReplica());
+
+        logger.info("{} Activating transfer {}, {} ms since pending", logPrefix(), this, currentTimeMillis() - createdAt);
+        ColumnFamilyStore cfs = ColumnFamilyStore.getIfExists(tableId);
+        Preconditions.checkNotNull(cfs);
+        Preconditions.checkState(!sstables.isEmpty());
+
+        if (activation.dryRun)
+        {
+            logger.info("{} Not adding SSTables to live set for dryRun {}", logPrefix(), activation);
+            return;
+        }
+
+        // Ensure no lingering mutation IDs, only activation IDs
+        for (SSTableReader sstable : sstables)
+        {
+            Preconditions.checkState(sstable.getCoordinatorLogOffsets().isEmpty());
+
+            // Modify SSTables metadata to durably set transfer ID before importing
+            ImmutableCoordinatorLogOffsets logOffsets = new ImmutableCoordinatorLogOffsets.Builder()
+                                                  .addTransfer(activation.activationId)
+                                                  .build();
+            try
+            {
+                sstable.mutateCoordinatorLogOffsetsAndReload(logOffsets);
+            }
+            catch (IOException e)
+            {
+                throw new RuntimeException(e);
+            }
+
+            Preconditions.checkState(sstable.getCoordinatorLogOffsets().isEmpty());
+            Preconditions.checkState(!sstable.getCoordinatorLogOffsets().transfers().isEmpty());
```

Comment:
this is a little weird from an api perspective. We're asserting that the log offsets are empty, then asserting that a collection contained in them is not empty. We should probably rename the ImmutableLogOffsets.isEmpty to something indicating that we want to know if the offsets are empty, or if the transfers are empty. There may also be other spots in the code where we take some action if the log offsets are empty without confirming that the transfers are also empty. So we should rethink some parts of this to make it less likely to be misused